File: ex4/politica_emprestimo.py
class PoliticaEmprestimo:

    PRAZOS = {
        "ALUNO": {
            "LIVRO": 14,
            "REVISTA": 7,
            "MIDIA": 3
        },
        "PROFESSOR": {
            "LIVRO": 30,
            "REVISTA": 14,
            "MIDIA": 7
        }
    }

    TAXAS_MULTA = {
        "LIVRO": 1.0,
        "REVISTA": 1.5,
        "MIDIA": 3.0
    }

    @staticmethod
    def _validar_perfil(perfil):
        if perfil not in PoliticaEmprestimo.PRAZOS:
            raise ValueError("Perfil inválido")

    @staticmethod
    def _validar_material(material):
        if material not in PoliticaEmprestimo.TAXAS_MULTA:
            raise ValueError("Material inválido")

    @staticmethod
    def prazo_emprestimo(perfil, material):
        PoliticaEmprestimo._validar_perfil(perfil)
        PoliticaEmprestimo._validar_material(material)

        return PoliticaEmprestimo.PRAZOS[perfil][material]

    @staticmethod
    def calcular_multa(material, dias_atraso):
        PoliticaEmprestimo._validar_material(material)

        if dias_atraso <= 0:
            return 0.0

        taxa = PoliticaEmprestimo.TAXAS_MULTA[material]

        dias_normais = min(dias_atraso, 10)
        dias_agravados = max(dias_atraso - 10, 0)

        return (
            dias_normais * taxa
            + dias_agravados * taxa * 2
        )

    @staticmethod
    def pode_renovar(
        perfil,
        material,
        dias_atraso,
        possui_reserva,
        renovacoes_realizadas
    ):
        PoliticaEmprestimo._validar_perfil(perfil)
        PoliticaEmprestimo._validar_material(material)

        if renovacoes_realizadas < 0:
            raise ValueError(
                "Número de renovações inválido"
            )

        if dias_atraso > 0 or possui_reserva:
            return False

        if material == "MIDIA":
            return False

        if material == "REVISTA":
            return renovacoes_realizadas < 1

        limite = 2 if perfil == "ALUNO" else 3

        return renovacoes_realizadas < limite

File: ex4/PoliticaEmprestimo.java
import java.util.Map;

public final class PoliticaEmprestimo {

    private static final Map<String, Map<String, Integer>> PRAZOS =
        Map.of(
            "ALUNO",
            Map.of(
                "LIVRO", 14,
                "REVISTA", 7,
                "MIDIA", 3
            ),

            "PROFESSOR",
            Map.of(
                "LIVRO", 30,
                "REVISTA", 14,
                "MIDIA", 7
            )
        );

    private static final Map<String, Double> TAXAS_MULTA =
        Map.of(
            "LIVRO", 1.0,
            "REVISTA", 1.5,
            "MIDIA", 3.0
        );

    private PoliticaEmprestimo() {
    }

    private static void validarPerfil(String perfil) {
        if (!PRAZOS.containsKey(perfil)) {
            throw new IllegalArgumentException(
                "Perfil inválido"
            );
        }
    }

    private static void validarMaterial(String material) {
        if (!TAXAS_MULTA.containsKey(material)) {
            throw new IllegalArgumentException(
                "Material inválido"
            );
        }
    }

    public static int prazoEmprestimo(
        String perfil,
        String material
    ) {
        validarPerfil(perfil);
        validarMaterial(material);

        return PRAZOS.get(perfil).get(material);
    }

    public static double calcularMulta(
        String material,
        int diasAtraso
    ) {
        validarMaterial(material);

        if (diasAtraso <= 0) {
            return 0.0;
        }

        double taxa = TAXAS_MULTA.get(material);

        int diasNormais =
            Math.min(diasAtraso, 10);

        int diasAgravados =
            Math.max(diasAtraso - 10, 0);

        return diasNormais * taxa
             + diasAgravados * taxa * 2;
    }

    public static boolean podeRenovar(
        String perfil,
        String material,
        int diasAtraso,
        boolean possuiReserva,
        int renovacoesRealizadas
    ) {
        validarPerfil(perfil);
        validarMaterial(material);

        if (renovacoesRealizadas < 0) {
            throw new IllegalArgumentException(
                "Número de renovações inválido"
            );
        }

        if (diasAtraso > 0 || possuiReserva) {
            return false;
        }

        if (material.equals("MIDIA")) {
            return false;
        }

        if (material.equals("REVISTA")) {
            return renovacoesRealizadas < 1;
        }

        int limite =
            perfil.equals("ALUNO") ? 2 : 3;

        return renovacoesRealizadas < limite;
    }
}